File: app/src/main/java/isaacborges/com/projetoconcrete/PullRequestActivity.java
package isaacborges.com.projetoconcrete;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.view.MenuItem;
import android.widget.Toast;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.List;

import isaacborges.com.projetoconcrete.eventbus.ExecutaServico;
import isaacborges.com.projetoconcrete.eventbus.events.BuscaListaPullRequestsEvent;
import isaacborges.com.projetoconcrete.model.PullRequest;
import isaacborges.com.projetoconcrete.model.Repositorio;
import isaacborges.com.projetoconcrete.recyclerview.adapters.ListaPullRequestAdapter;
import isaacborges.com.projetoconcrete.recyclerview.adapters.listeners.OnPullRequestItemClickListener;
import isaacborges.com.projetoconcrete.utils.ConversorJson;

import static isaacborges.com.projetoconcrete.ConstantesDaAplicacao.CHAVE_PULLREQUESTS;
import static isaacborges.com.projetoconcrete.ConstantesDaAplicacao.CHAVE_REPOSITORIO;

public class PullRequestActivity extends AppCompatActivity {

    private RecyclerView recyclerViewPullRequests;

    private ListaPullRequestAdapter adapter;
    private Repositorio repositorio;

    private final EventBus eventBus = EventBus.getDefault();;
    private final ExecutaServico executaServico = new ExecutaServico(this);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_pull_request);

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        buscaReferenciasDasViews();
        configuraAdapter();
        configuraRecyclerView();
        carregaRepositorioRecebido();

        if(savedInstanceState != null){
            buscaDadosDoBundle(savedInstanceState);
        }else{
            buscaPullRequests();

        }
    }

    private void configuraRecyclerView() {
        recyclerViewPullRequests.setAdapter(adapter);
    }

    private void buscaDadosDoBundle(Bundle savedInstanceState) {
        String jsonInString = savedInstanceState.getString(CHAVE_PULLREQUESTS);
        List<PullRequest> listaPullRequests = (List<PullRequest>) ConversorJson.converteJsonParaLista(jsonInString, PullRequest.class);
        adapter.atualiza(listaPullRequests);
    }

    private void buscaPullRequests() {
        String nomeAutor = repositorio.getAutor().getLogin();
        String nomeRepositorio = repositorio.getName();
        executaServico.buscaPullRequests(nomeAutor, nomeRepositorio, new ExecutaServico.ErroAoCarregarInformacoes() {
            @Override
            public void erroAoCarregar(String mensagem) {
                Toast.makeText(PullRequestActivity.this, mensagem, Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void atualizaListaAlunoEvent(BuscaListaPullRequestsEvent event){
        adapter.atualiza(event.getPullRequests());
    }

    @Override
    public void onStart() {
        super.onStart();
        eventBus.register(this);
    }

    @Override
    public void onStop() {
        super.onStop();
        eventBus.unregister(this);
    }

    private void carregaRepositorioRecebido() {
        Intent intent = getIntent();
        if(intent.hasExtra(CHAVE_REPOSITORIO)){
            repositorio = intent.getParcelableExtra(CHAVE_REPOSITORIO);
            setTitle(repositorio.getName());
        }
    }

    private void buscaReferenciasDasViews() {
        recyclerViewPullRequests = findViewById(R.id.recyclerViewPullRequest);
    }

    private void configuraAdapter() {
        adapter = new ListaPullRequestAdapter(this);
        adapter.setOnPullRequestItemClickListener(new OnPullRequestItemClickListener() {
            @Override
            public void onItemClick(PullRequest pullRequest, int posicao) {
                abreSiteDoPullRequest(pullRequest);
            }
        });
    }

    private void abreSiteDoPullRequest(PullRequest pullRequest) {
        Intent intentSite = new Intent(Intent.ACTION_VIEW); //Intent implícita
        intentSite.setData(Uri.parse(pullRequest.getHtml_url()));
        startActivity(intentSite);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        String jsonString = ConversorJson.converteListaParaJson(adapter.getListaPullRequests());
        outState.putString(CHAVE_PULLREQUESTS, jsonString);
        super.onSaveInstanceState(outState);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/br/com/rmso/popularrepositories/ui/pullrequest/PullRequestActivity.kt
package br.com.rmso.popularrepositories.ui.pullrequest

import android.content.Intent
import android.net.Uri
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import br.com.rmso.popularrepositories.utils.ListOnClickListener
import br.com.rmso.popularrepositories.R
import br.com.rmso.popularrepositories.model.PullRequest
import br.com.rmso.popularrepositories.utils.Constants
import kotlinx.android.synthetic.main.activity_pull_request.*

class PullRequestActivity : AppCompatActivity(),
    ListOnClickListener, IPullRequest.View{

    private var owner = ""
    private var repositoryName = ""
    private var pullRequestArrayList = ArrayList<PullRequest>()
    private val constans = Constants()
    var linearLayoutManager = LinearLayoutManager(this@PullRequestActivity)
    var adapterRepository = PullResquestAdapter(
        pullRequestArrayList,
        this@PullRequestActivity
    )
    private var pullRequestPresenter: IPullRequest.Presenter = PullRequestPresenter(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_pull_request)

        owner = intent.getStringExtra(constans.owner)
        repositoryName = intent.getStringExtra(constans.repository)

        toolbar.title = repositoryName
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setDisplayShowHomeEnabled(true)

        if (savedInstanceState != null) {
            pullRequestArrayList.addAll(savedInstanceState.getParcelableArrayList(constans.listPullRequestInstance))
            owner = savedInstanceState.getString(constans.owner)
            repositoryName = savedInstanceState.getString(constans.repository)
        }else {
            (pullRequestPresenter as PullRequestPresenter).requestList(pullRequestArrayList, owner, repositoryName)
        }

        rv_pull_request.apply {
            layoutManager = linearLayoutManager
            adapter = adapterRepository
        }
    }

    override fun updateList(list: List<PullRequest>?, owner: String, repositoryName: String) {
        pullRequestArrayList.addAll(list!!)
        rv_pull_request.adapter?.notifyDataSetChanged()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelableArrayList(constans.listPullRequestInstance, pullRequestArrayList)
        outState.putString(constans.owner, owner)
        outState.putString(constans.repository, repositoryName)
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return super.onSupportNavigateUp()
    }

    override fun onClick(position: Int) {
        val pullRequest = pullRequestArrayList[position]
        val url = pullRequest.html_url
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
        startActivity(intent)

    }

    override fun progressBar(status: Boolean) {
        if(status) {
            pb_loading_pull.visibility = View.VISIBLE
        }else {
            pb_loading_pull.visibility = View.GONE
        }
    }

    override fun errorRequest(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show()
    }

}
